cv-data-controller/src/main/java/com/trihydro/cvdatacontroller/controller/CdotUpstreamPathController.java

```diff
@@ -0,0 +1,222 @@
+    for (int i = 0; i < mileposts.size(); i++) {
+      Milepost currentMilepost = mileposts.get(i);
+      BigDecimal currentLatitude = currentMilepost.getLatitude().setScale(14, BigDecimal.ROUND_HALF_UP);
+      BigDecimal currentLongitude = currentMilepost.getLongitude().setScale(14, BigDecimal.ROUND_HALF_UP);
+      if (latitude.equals(currentLatitude) && longitude.equals(currentLongitude)) {
+        return i;
+      }
+    }
+    return -1;
+  }
+
+  private String convertMilepostsToGeojsonString(List<Milepost> mileposts) {
+    StringBuilder geojsonStringBuilder = new StringBuilder();
+    geojsonStringBuilder.append("{ \"type\": \"FeatureCollection\", \"features\": [{ \"type\": \"Feature\", \"geometry\": { \"type\": \"LineString\", \"coordinates\": [");
+    for (int i = 0; i < mileposts.size(); i++) {
+      Milepost milepost = mileposts.get(i);
+      geojsonStringBuilder.append("[");
+      geojsonStringBuilder.append(milepost.getLongitude().toString());
+      geojsonStringBuilder.append(", ");
+      geojsonStringBuilder.append(milepost.getLatitude().toString());
+      geojsonStringBuilder.append("]");
+      if (i < mileposts.size() - 1) {
+        geojsonStringBuilder.append(", ");
+      }
+    }
+    geojsonStringBuilder.append("] }, \"properties\": { \"commonName\": \"");
+    geojsonStringBuilder.append(mileposts.get(0).getCommonName());
+    geojsonStringBuilder.append("\" } }] }");
+    return geojsonStringBuilder.toString();
+  }
+
+  // define path direction enum (ASCENDING, DESCENDING)
+  public enum PathDirection {
+    ASCENDING,
+    DESCENDING
+  }
+
+  /**
+   * Helper class to calculate distance between two points and total distance of a buffer path
+   */
+  public static class DistanceCalculator {
+    public static double calculateDistanceInMiles(List<Milepost> buffer) {
+      double distanceInMiles = 0;
+      for (int i = 0; i < buffer.size() - 1; i++) {
+        Milepost mp1 = buffer.get(i);
+        Milepost mp2 = buffer.get(i + 1);
+        double distanceInMeters = DistanceCalculator.calculateDistanceInMetersBetweenTwoPoints(mp1.getLatitude().doubleValue(),
+            mp1.getLongitude().doubleValue(), mp2.getLatitude().doubleValue(), mp2.getLongitude().doubleValue());
+        distanceInMiles += distanceInMeters / 1609.34;
+      }
+      return distanceInMiles;
+    }
+
+    public static double calculateDistanceInMetersBetweenTwoPoints(double lat1, double lon1, double lat2, double lon2) {
+      final int R = 6371000; // Radius of the earth in meters
+      double latDistance = Math.toRadians(lat2 - lat1);
+      double lonDistance = Math.toRadians(lon2 - lon1);
+      double a = Math.sin(latDistance / 2) * Math.sin(latDistance / 2)
+          + Math.cos(Math.toRadians(lat1)) * Math.cos(Math.toRadians(lat2))
+          * Math.sin(lonDistance / 2) * Math.sin(lonDistance / 2);
+      double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
+      return R * c; // Distance in meters
+    }
+  }
```

Comment:
suggestion: there is similar logic to these calculations in https://github.com/CDOT-CV/TIM-Manager/blob/6ea1b03af792e5f36690c10407002933536a609b/cv-data-service-library/src/main/java/com/trihydro/library/model/Milepost.java#L124. We should consider moving the logic over into that class to keep logic co-located.

Side-note: we can use memorization to remove the need to calculate the distance in miles without the use of a list. we can achieve that by calculating the distance between two points, storing that result, and adding the result to later calculations. that will improve performance and allow us to move these methods to the Milepost class (we'd only need two Mileposts to calculate distance)

Reply:
I love the idea of using memoization here. Do you think it would be worth it to save the data in a database table or just should we just keep it in memory?

Reply:
no need for a db table. in memory should be good in this case

Reply:
I've updated the DistanceCalculator to use the Milepost.angularDistanceTo() method. While I attempted to refactor the code to eliminate the double for-loop, those changes caused unit tests to fail. After spending significant time on this, I recommend adding optimization to the backlog. Additionally, I've included JavaDocs to document the inefficiency of recalculating the buffer path during each iteration.